File: app/src/main/java/ru/buylist/data/db/BuyListDatabase.java
package ru.buylist.data.db;


import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;

import java.util.List;

import ru.buylist.R;
import ru.buylist.data.dao.*;
import ru.buylist.data.entity.*;
import ru.buylist.utils.AppExecutors;
import ru.buylist.utils.CategoryDatabaseWorker;

@Database(entities = {Collection.class, Item.class, GlobalItem.class, Category.class}, version = 1, exportSchema = false)
public abstract class BuyListDatabase extends RoomDatabase {

    public static final String DATABASE_NAME = "buyListBase";
    private static BuyListDatabase instance;

    public static BuyListDatabase getInstance(final Context context, final AppExecutors executors) {
        if (instance == null) {
            synchronized (BuyListDatabase.class) {
                if (instance == null) {
                    instance = Room.databaseBuilder(context.getApplicationContext(), BuyListDatabase.class,
                            DATABASE_NAME)
                            .addCallback(new Callback() {
                                @Override
                                public void onCreate(@NonNull SupportSQLiteDatabase db) {
                                    super.onCreate(db);
                                    executors.discIO().execute(() -> {
                                        List<Category> categories = CategoryDatabaseWorker.getStandartCategories(context);
                                        for (Category category : categories) {
                                            getInstance(context, executors).categoryDao().addCategory(category);
                                        }
                                    });
                                }
                            })
                            .build();
                }
            }
        }
        return instance;
    }

    public abstract CollectionDao collectionDao();

    public abstract ItemDao itemDao();

    public abstract GlobalItemDao globalItemDao();

    public abstract CategoryDao categoryDao();
}

File: app/src/main/java/ru/buylist/data/db/BuyListDatabase.kt
package ru.buylist.data.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import ru.buylist.data.dao.*
import ru.buylist.data.entity.*
import ru.buylist.utils.ProductsWorker

@Database(entities = [BuyList::class, Pattern::class, Recipe::class, GlobalItem::class],
        version = 1, exportSchema = false)
abstract class BuyListDatabase : RoomDatabase() {
    abstract fun buyListDao(): BuyListDao
    abstract fun patternDao(): PatternDao
    abstract fun recipeDao(): RecipeDao
    abstract fun globalItemDao(): GlobalItemDao

    companion object {
        private const val DATABASE_NAME = "buyListBase"

        @Volatile private var instance: BuyListDatabase? = null

        fun getInstance(context: Context): BuyListDatabase {
            return instance ?: synchronized(this) {
                instance
                        ?: buildDatabase(context).also { instance = it }
            }
        }

        private fun buildDatabase(context: Context): BuyListDatabase {
            return Room.databaseBuilder(context, BuyListDatabase::class.java, DATABASE_NAME)
                    .addCallback(object : RoomDatabase.Callback() {
                        override fun onCreate(db: SupportSQLiteDatabase) {
                            super.onCreate(db)
                            val request = OneTimeWorkRequestBuilder<ProductsWorker>().build()
                            WorkManager.getInstance(context).enqueue(request)
                        }
                    })
                    .build()
        }
    }
}